File: kg_core/response.py
#  Copyright 2022 EBRAINS AISBL
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0.
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
#  This open source software code was developed in part or in whole in the
#  Human Brain Project, funded from the European Union's Horizon 2020
#  Framework Programme for Research and Innovation under
#  Specific Grant Agreements No. 720270, No. 785907, and No. 945539
#  (Human Brain Project SGA1, SGA2 and SGA3).
#

from __future__ import annotations
import http.client
import uuid
from abc import ABC
from enum import Enum, EnumMeta
from typing import Any, Callable, Iterable, Optional, Dict, TypeVar, Generic, List
from uuid import UUID

from pydantic import BaseModel, Field

from kg_core.__communication import KGRequestWithResponseContext


class ReleaseStatus(str, Enum):
    RELEASED = "RELEASED"
    UNRELEASED = "UNRELEASED"
    HAS_CHANGED = "HAS_CHANGED"


class JsonLdDocument(Dict[str, Any]):
    def __init__(self, seq: Iterable[List[str]] = (), id_namespace: Optional[str] = None, **kwargs: Any):
        super(JsonLdDocument, self).__init__(seq, **kwargs)
        self._id_namespace = id_namespace

    def to_uuid(self, value) -> Optional[UUID]:
        if value and value.startswith(self._id_namespace):
            return uuid.UUID(value[len(self._id_namespace):])
        else:
            return None


class Instance(JsonLdDocument):
    uuid: Optional[UUID] = None

    def __init__(self, seq: Iterable[List[str]] = (), id_namespace: Optional[str] = None, **kwargs: Any):
        super(Instance, self).__init__(seq, id_namespace, **kwargs)
        self.instance_id = self["@id"] if "@id" in self else None
        self.uuid = self.to_uuid(self.instance_id)

    def __str__(self):
        return f"Instance {self.uuid if self.uuid else 'unknown'}"


class TermsOfUse(BaseModel):
    accepted: bool = False
    version: str
    data: str


class Error(BaseModel):
    code: int
    message: Optional[str] = None
    uuid: Optional[UUID] = Field(None, alias="instanceId")


class Scope(BaseModel):
    uuid: Optional[UUID] = Field(None, alias="id")
    label: Optional[str] = None
    space: Optional[str] = None
    types: Optional[List[str]] = None
    children: Optional[List[Scope]] = None
    permissions: Optional[List[str]] = None


class SpaceInformation(BaseModel):
    identifier: Optional[str] = Field(None, alias="http://schema.org/identifier")
    name: Optional[str] = Field(None, alias="http://schema.org/name")
    permissions: Optional[List[str]] = Field(None, alias="https://core.kg.ebrains.eu/vocab/meta/permissions")


class TypeInformation(BaseModel):
    identifier: Optional[str] = Field(None, alias="http://schema.org/identifier")
    description: Optional[str] = Field(None, alias="http://schema.org/description")
    name: Optional[str] = Field(None, alias="http://schema.org/name")
    # TODO incoming_links
    occurrences: Optional[int] = Field(None, alias="https://core.kg.ebrains.eu/vocab/meta/occurrences")

    # TODO properties
    # TODO spaces


class ReducedUserInformation(BaseModel):
    alternate_name: Optional[str] = Field(None, alias="http://schema.org/alternateName")
    name: Optional[str] = Field(None, alias="http://schema.org/name")
    uuid: Optional[UUID] = Field(None, alias="@id")


class ListOfUUID(List[UUID]):
    def __init__(self, seq: Iterable[UUID] = ()):
        super(ListOfUUID, self).__init__([UUID(s) for s in seq])


class ListOfReducedUserInformation(List[ReducedUserInformation]):
    def __init__(self, seq: Iterable[ReducedUserInformation] = ()):
        super(ListOfReducedUserInformation, self).__init__([ReducedUserInformation(**s) for s in seq])


class User(BaseModel):
    alternate_name: Optional[str] = Field(None, alias="http://schema.org/alternateName")
    name: Optional[str] = Field(None, alias="http://schema.org/name")
    email: Optional[str] = Field(None, alias="http://schema.org/email")
    given_name: Optional[str] = Field(None, alias="http://schema.org/givenName")
    family_name: Optional[str] = Field(None, alias="http://schema.org/familyName")
    identifiers: Optional[List[str]] = Field(None, alias="http://schema.org/identifier")


class UserWithRoles(BaseModel):
    user: User
    client_roles: Optional[List[str]] = Field(None, alias="clientRoles")
    user_roles: Optional[List[str]] = Field(None, alias="userRoles")
    invitations: Optional[List[str]] = None
    client_id: Optional[str] = Field(None, alias="clientId")

    # permissions


ResponseType = TypeVar("ResponseType")


def translate_error(response: KGRequestWithResponseContext) -> Optional[Error]:
    if response.content and "error" in response.content and response.content["error"] and type(response.content["error"]) != str:
        return Error(**response.content["error"])
    else:
        return Error(code=response.status_code, message=http.client.responses[response.status_code]) if response.status_code and response.status_code >= 400 else None


class _AbstractResult(ABC):

    def __init__(self, response: KGRequestWithResponseContext):
        self.message: Optional[str] = response.content[
            "message"] if response.content and "message" in response.content else None
        self.start_time: Optional[int] = response.content[
            "startTime"] if response.content and "startTime" in response.content else None
        self.duration_in_ms: Optional[int] = response.content[
            "durationInMs"] if response.content and "durationInMs" in response.content else None
        self.transaction_id: Optional[int] = response.content[
            "transactionId"] if response.content and "transactionId" in response.content else None
        self.error: Optional[Error] = translate_error(response)


class _AbstractResultPage(_AbstractResult):
    def __init__(self, response: KGRequestWithResponseContext):
        super(_AbstractResultPage, self).__init__(response)
        self.total: Optional[int] = response.content[
            "total"] if response.content and "total" in response.content else None
        self.size: Optional[int] = response.content["size"] if response.content and "size" in response.content else None
        self.start_from: Optional[int] = response.content[
            "from"] if response.content and "from" in response.content else None


class ResponseObjectConstructor(Generic[ResponseType]):
    @staticmethod
    def init_response_object(constructor: Callable[..., ResponseType], data: Any, id_namespace: Any) -> ResponseType:
        if issubclass(constructor, BaseModel):
            return constructor(**data)
        elif issubclass(constructor, Enum):
            # Not pretty but works for now
            return constructor[data]  # type: ignore
        elif constructor == JsonLdDocument or constructor == Instance:
            return constructor(data, id_namespace)
        else:
            return constructor(data)


class ResultPageIterator(Generic[ResponseType]):

    def __init__(self, result_page: ResultPage[ResponseType]):
        self._result_page: ResultPage[ResponseType] = result_page

    def __iter__(self):
        self.n = 0
        return self

    def __next__(self) -> Optional[ResultPage[ResponseType]]:
        if self._result_page:
            if self._result_page.error:
                raise ValueError(self._result_page.error.message)
            elif self._result_page.data:
                if self._result_page.total is None or (self._result_page.total and self.n < self._result_page.total):
                    if self.n >= self._result_page.start_from + self._result_page.size and (
                            self._result_page.has_next_page() is None or self._result_page.has_next_page()):
                        self._result_page = self._result_page.next_page()
                    if self._result_page:
                        result = self._result_page.data[self.n - self._result_page.start_from]
                        self.n += 1
                        return result
        raise StopIteration


class ResultPage(_AbstractResultPage, Generic[ResponseType]):

    def __init__(self, response: KGRequestWithResponseContext, constructor: Callable[..., ResponseType]):
        super(ResultPage, self).__init__(response)
        self.data: Optional[List[ResponseType]] = [
            ResponseObjectConstructor.init_response_object(constructor, r, response.id_namespace) for r in
            response.content["data"]] if response.content and "data" in response.content else None
        self._original_response = response
        self._original_constructor = constructor

    def __str__(self):
        return f"{super.__str__(self)} - status: {self.error.code if self.error else 'success'}"

    def next_page(self) -> Optional[ResultPage[ResponseType]]:
        """ returns the next page of this result if there is one - otherwise returns None """
        next_page = self.has_next_page()
        if next_page is None or next_page:  # next page can be
            result = self._original_response.next_page(self.start_from, self.size)
            result_page = ResultPage[ResponseType](response=result,
                                                   constructor=self._original_constructor) if result else None
            if result_page and result_page.data:
                return result_page
            else:
                return None
        return None

    def has_next_page(self) -> Optional[bool]:
        """ returns True if a next page exists. Returns None if the original request has been executed without "full count" (by setting the "returnTotalResults" to false). """
        if self.total:
            if self.total is not None and self.start_from is not None and self.size is not None:
                return self.start_from + self.size < self.total
            return False
        return None

    def items(self) -> ResultPageIterator[ResponseType]:
        """ returns an iterator to be used e.g. within a for loop. Attention: Do not manipulate the underlying data structure within the loop! The resolution of pages is lazy and manipulations while iterating can lead to unexpected results."""
        return ResultPageIterator(self)


class Result(_AbstractResult, Generic[ResponseType]):

    def __init__(self, response: KGRequestWithResponseContext, constructor: Callable[..., ResponseType]):
        super(Result, self).__init__(response)
        self.data: Optional[ResponseType] = ResponseObjectConstructor.init_response_object(constructor,
                                                                                           response.content["data"],
                                                                                           response.id_namespace) if response.content and "data" in response.content and \
                                                                                                                     response.content[
                                                                                                                         "data"] is not None else None

    def __str__(self):
        return f"{super.__str__(self)} - status: {str(self.error.code) + ' (' + self.error.message + ')' if self.error is not None else 'success'}"


class ResultsById(_AbstractResult, Generic[ResponseType]):

    def __init__(self, response: KGRequestWithResponseContext, constructor: Callable[..., ResponseType]):
        super(ResultsById, self).__init__(response)
        self.data: Optional[Dict[str, Result[ResponseType]]] = {
            k: Result[ResponseType](response.copy_context(r), constructor) for k, r in
            response.content["data"].items()} if response.content and "data" in response.content and response.content[
            "data"] else None

    def __str__(self):
        return f"{super.__str__(self)} - status: {self.error.code if self.error else 'success'}"

File: kgCoreJs/response.ts
/*  Copyright 2022 EBRAINS AISBL
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0.
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  This open source software code was developed in part or in whole in the
 *  Human Brain Project, funded from the European Union's Horizon 2020
 *  Framework Programme for Research and Innovation under
 *  Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 *  (Human Brain Project SGA1, SGA2 and SGA3).
 */
import { KGRequestWithResponseContext } from "./communication";

type UUID = string;

export enum ReleaseStatus {
  RELEASED = "RELEASED",
  UNRELEASED = "UNRELEASED",
  HAS_CHANGED = "HAS_CHANGED",
}
export class JsonLdDocument {
  [index: string]: any;
  idNamespace?: string;
  constructor(json: any, idNamespace?: string) {
    Object.keys(json).forEach((key: string) => {
      this[key] = json[key];
    });
    this.idNamespace = idNamespace;
  }

  toUuid(value: string): UUID | null {
    if (this.idNamespace) {
      if (value && value.startsWith(this.idNamespace)) {
        const r = value.split("/");
        return r[r.length - 1];
      }
    }
    return null;
  }
}

export class Instance extends JsonLdDocument {
  uuid: UUID | null;
  constructor(data: any, idNamespace?: string) {
    super(data, idNamespace);
    this.instanceId = this["@id"] ?? null;
    this.uuid = this.toUuid(this.instanceId);
  }
}

export class TermsOfUse {
  accepted: boolean;
  version: string;
  data: string;
  constructor(data: any) {
    this.accepted = data["accepted"] ?? false;
    this.version = data["version"];
    this.data = data["data"];
  }
}

export class KGError {
  code?: number;
  message?: string;
  uuid?: UUID;
  constructor(
    code?: number,
    message?: string,
    uuid?: UUID
  ) {
    this.code = code;
    this.message = message;
    this.uuid = uuid;
  }
}

export class Scope {
  uuid?: UUID;
  label?: string;
  space?: string;
  types?: Array<string>;
  children?: Array<Scope>;
  permissions?: Array<string>;
  constructor(data: any) {
    this.uuid = data["id"];
    this.label = data["label"];
    this.space = data["space"];
    this.types = data["types"];
    this.children = data["children"];
    this.permissions = data["permissions"];
  }
}

export class SpaceInformation {
  identifier?: string;
  name?: string;
  permissions?: Array<string>;
  constructor(data: any) {
    this.identifier = data["http://schema.org/identifier"];
    this.name = data["http://schema.org/name"];
    this.permissions = data["https://core.kg.ebrains.eu/vocab/meta/permissions"];
  }
}

export class TypeInformation {
  identifier?: string;
  description?: string;
  name?: string;
  occurrences?: number;
  constructor(data: any) {
    this.identifier = data["http://schema.org/identifier"];
    this.description = data["http://schema.org/description"];
    this.name = data["http://schema.org/name"];
    this.occurrences = data["https://core.kg.ebrains.eu/vocab/meta/occurrences"];
  }
}

class ReducedUserInformation {
  alternateName?: string;
  name?: string;
  uuid?: UUID;
  constructor(data: any) {
    this.alternateName = data["http://schema.org/alternateName"];
    this.name = data["http://schema.org/name"];
    this.uuid = data["@id"];
  }
}

export class ListOfUUID extends Array<string> {
  //WARNING: Do not extend this class
  constructor(items: Array<string>) {
    super(...items);
    Object.setPrototypeOf(this, Array.prototype);
  }
}
export class ListOfReducedUserInformation extends Array<ReducedUserInformation> {
  //WARNING: Do not extend this class
  constructor(items: Array<ReducedUserInformation>) {
    super(...items);
    Object.setPrototypeOf(this, Array.prototype);
  }
}
export class User {
  alternateName?: string;
  name?: string;
  email?: string;
  givenName?: string;
  familyName?: string;
  identifiers?: Array<string>;
  constructor(user: any) {
    this.alternateName = user["http://schema.org/alternateName"];
    this.name = user["http://schema.org/name"];
    this.email = user["http://schema.org/email"];
    this.givenName = user["http://schema.org/givenName"];
    this.familyName = user["http://schema.org/familyName"];
    this.identifiers = user["http://schema.org/identifier"];
  }
}

class UserWithRoles {
  user: User;
  clientRoles?: Array<string>;
  userRoles?: Array<string>;
  invitations?: Array<string>;
  clientId?: string;
  constructor(
    user: User,
    clientRoles?: Array<string>,
    userRoles?: Array<string>,
    invitations?: Array<string>,
    clientId?: string
  ) {
    this.user = user;
    this.clientRoles = clientRoles;
    this.userRoles = userRoles;
    this.invitations = invitations;
    this.clientId = clientId;
  }
}

export declare class ErrorStatusText {
  [index: number]: string;
}

const errorStatusText: ErrorStatusText = {
  400: "Bad Request",
  401: "Unauthorized",
  403: "Forbidden",
  404: "Not Found",
  405: "Method Not Allowed",
  406: "Not Acceptable",
  407: "Proxy Authentication Required",
  408: "Request Timeout",
  409: "Conflict",
  410: "Gone",
  411: "Length Required",
  412: "Precondition Failed",
  413: "Payload Too Large",
  414: "URI Too Long",
  415: "Unsupported Media Type",
  416: "Range Not Satisfiable",
  417: "Expectation Failed",
  421: "Misdirected Request",
  425: "Too Early",
  426: "Upgrade Required",
  428: "Precondition Required",
  429: "Too Many Requests",
  431: "Request Header Fields Too Large",
  451: "Unavailable For Legal Reasons",
  500: "Internal Server Error",
  501: "Not Implemented",
  502: "Bad Gateway",
  503: "Service Unavailable",
  504: "Gateway Timeout",
  505: "HTTP Version Not Supported",
  506: "Variant Also Negotiates",
  510: "Not Extended",
  511: "Network Authentication Required",
  420: "Method Failure",
  598: "Network read timeout",
  599: "Network Connect Timeout",
  440: "Login Time-out",
  444: "No Response",
  494: "Request header too large",
  495: "SSL Certificate Error",
  496: "SSL Certificate Required",
  497: "HTTP Request Sent to HTTPS Port",
  499: "Client Closed Request"
};

export const translateError = (response: KGRequestWithResponseContext) => {
  if (
    response?.content?.error &&
    !(response.content.error instanceof String)
  ) {
    return new KGError(
      response.statusCode,
      response.content.error,
      response.idNamespace
    );
  } else {
    if (response.statusCode && response.statusCode >= 400) {
      return new KGError(response.statusCode, errorStatusText[response.statusCode]);
    }
  }
  return null;
};

abstract class _AbstractResult {
  message?: string;
  startTime?: number;
  durationInMs?: number;
  transactionId?: number;
  error: KGError | null;
  constructor(response: KGRequestWithResponseContext) {
    this.message = response?.content?.message;
    this.startTime = response?.content?.startTime;
    this.durationInMs = response?.content?.durationInMs;
    this.transactionId = response?.content?.transactionId;
    this.error = translateError(response);
  }
}

class _AbstractResultPage extends _AbstractResult {
  total?: number;
  size?: number;
  startFrom?: number;
  constructor(response: KGRequestWithResponseContext) {
    super(response);
    this.total = response?.content?.total;
    this.size = response?.content?.size;
    this.startFrom = response?.content?.from;
  }
}

class ResponseObjectConstructor {
  static initResponseObject(constructor: any, data: any, idNamespace: any) {
    if (constructor === JsonLdDocument || constructor === Instance) {
      return new constructor(data, idNamespace);
    } else if (constructor === ReleaseStatus) {
      return constructor[data];
    }
    return new constructor(data);
  }
}

class ResultPageIterator<T> implements Iterator<T> {
  resultPage: ResultPage<T> | null;
  private counter: number = 0;
  constructor(resultPage: ResultPage<T> | null) {
    this.resultPage = resultPage;
  }

  next(): IteratorResult<T> {
    if (this.resultPage) {
      if (this.resultPage.error) {
        throw new Error(this.resultPage?.error?.message ?? undefined);
      } else if (this.resultPage.data) {
        if (
          !this.resultPage.total ||
          (this.resultPage.total && this.counter < this.resultPage.total)
        ) {
          if (this.resultPage.startFrom && this.resultPage.size) {
            if (
              this.counter >=
                this.resultPage.startFrom + this.resultPage.size &&
              this.resultPage.hasNextPage()
            ) {
              this.resultPage = this.resultPage.nextPage();
            }
          }
          if (this.resultPage && this.resultPage.startFrom) {
            const result =
              this.resultPage.data[this.counter - this.resultPage.startFrom];
            this.counter++;
            return { value: result, done: false };
          }
        }
      }
    }
    return { value: null, done: true };
  }
}

export class ResultPage<T> extends _AbstractResultPage {
  data: Array<T>;
  _originalResponse: any;
  _originalConstructor: any;
  constructor(response: KGRequestWithResponseContext, constructor: any) {
    super(response);
    this.data = this._getData(
      constructor,
      response.content,
      response.idNamespace
    );
    this._originalResponse = response;
    this._originalConstructor = constructor;
  }

  _getData(constructor:any, content: any, idNamespace:string) {
    if (content?.data) {
      const d: Array<T> = content.data;
      return d.map(c =>
        ResponseObjectConstructor.initResponseObject(
          constructor,
          c,
          idNamespace
        )
      );
    }
    return [];
  }

  nextPage(): ResultPage<T> | null {
    const nextPage = this.hasNextPage();
    if (nextPage === null || nextPage === true) {
      const result = this._originalResponse.nextPage(this.startFrom, this.size);
      const resultPage = new ResultPage<T>(result, this._originalConstructor);

      if (resultPage && resultPage.data) {
        return resultPage;
      }
      return null;
    }
    return null;
  }

  hasNextPage(): boolean | null {
    if (this.total) {
      if (this.startFrom && this.size) {
        return this.startFrom + this.size < this.total;
      }
      return false;
    }
    return null;
  }

  items() {
    return new ResultPageIterator<T>(this);
  }
}

export class Result<T> extends _AbstractResult {
  data: T | null;
  constructor(response: KGRequestWithResponseContext, constructor: any) {
    super(response);
    this.data = response?.content?.data
      ? ResponseObjectConstructor.initResponseObject(
          constructor,
          response.content?.data,
          response.idNamespace
        )
      : null;
  }
}

interface ResultById<T> {
  [index:string]: Result<T>
};

export class ResultsById<T> extends _AbstractResult {
  data: ResultById<T>|null;
  constructor(response: KGRequestWithResponseContext, constructor: any) {
    super(response);
    this.data = response?.content?.data
      ? this._getData(response.content?.data, response, constructor)
      : null;
  }

  _getData(data:any, response:KGRequestWithResponseContext, constructor: any) {
    return Object.entries(data).reduce((newObj, [key, val]) => {
      newObj[key] = new Result<T>(response.copyContext(val), constructor);
      return newObj;
    }, {} as ResultById<T>);
  }
}
